core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -43,7 +54,132 @@ impl RequestorNegotiationEngine {
+            Err(QueryEventsError::InvalidMaxEvents(max_events))?
+        } else if max_events == 0 {
+            return Ok(vec![]);
+        }
+
+        loop {
+            let events = get_events_from_db(&self.db, subscription_id, max_events).await?;
+            if events.len() > 0 {
+                return Ok(events);
+            }
+
+            // Solves panic 'supplied instant is later than self'.
+            if stop_time < Instant::now() {
+                return Ok(vec![]);
+            }
+            timeout = stop_time - Instant::now();
+
+            if let Err(error) = self
+                .notifier
+                .wait_for_event_with_timeout(subscription_id, timeout)
+                .await
+            {
+                return match error {
+                    NotifierError::Timeout(_) => Ok(vec![]),
+                    NotifierError::ChannelClosed(_) => {
+                        Err(QueryEventsError::InternalError(format!("{}", error)))
+                    }
+                    NotifierError::Unsubscribed(id) => Err(QueryEventsError::Unsubscribed(id)),
+                };
+            }
+            // Ok result means, that event with required subscription id was added.
+            // We can go to next loop to get this event from db. But still we aren't sure
+            // that list won't be empty, because other query_events calls can wait for the same event.
+        }
+    }
+}
+
+async fn get_events_from_db(
+    db: &DbExecutor,
+    subscription_id: &SubscriptionId,
+    max_events: i32,
+) -> Result<Vec<RequestorEvent>, QueryEventsError> {
+    let events = db
+        .as_dao::<EventsDao>()
+        .take_requestor_events(subscription_id, max_events)
+        .await?;
+
+    // Map model events to client RequestorEvent.
+    let results = futures::stream::iter(events)
+        .then(|event| event.into_client_requestor_event(&db))
+        .collect::<Vec<Result<RequestorEvent, EventError>>>()
+        .await;
+
+    // Filter errors. Can we do something better with errors, than logging them?
+    Ok(results
+        .into_iter()
+        .inspect(|result| {
+            if let Err(error) = result {
+                log::warn!("Error converting event to client type: {}", error);
+            }
+        })
+        .filter_map(|event| event.ok())
+        .collect::<Vec<RequestorEvent>>())
+}
+
+pub async fn proposal_receiver_thread(
+    db: DbExecutor,
+    mut proposal_receiver: UnboundedReceiver<DraftProposal>,
+    notifier: EventNotifier,
+) {
+    while let Some(proposal) = proposal_receiver.recv().await {
+        let db = db.clone();
+        let notifier = notifier.clone();
+        match async move {
+            log::info!("Received proposal from matcher. Adding to events queue.");
+
+            // Add proposal to database together with Negotiation record.
+            let proposal = db
+                .as_dao::<ProposalDao>()
+                .new_initial_proposal(proposal.demand, proposal.offer)
```

Comment:
If we have DraftProposal, why should we split it here and not use it as argument type here?

Reply:
This way Dao uses only model types not strange types from outside.